### PR TITLE
Bolt12/diffusion test dns recoverability

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

**Comment by bolt12** [comment]
bors merge

**Review by coot — COMMENTED** [review]
This looks good.   The only important question is if we should make the `verify` function check the time, see the comment in the `Added DNS Recoverability test` patch.

File: ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs [hunk 1]
```diff
@@ -265,6 +273,115 @@ prop_diffusion_nolivelock defaultBearerInfo diffScript@(DiffusionScript l) =
...
+                        recovered t evs
+        DiffusionPublicRootPeerTrace (TracePublicRootFailure dns err) ->
+            let ttl = ttlForDnsError err 0
+             in verify (Map.singleton dns (addTime ttl t) <> toRecover)
+                        recovered t evs
+        DiffusionLocalRootPeerTrace (TraceLocalRootResult dap _) ->
+          let dns = dapDomain dap
+           in case Map.lookup dns toRecover of
+                Nothing -> verify toRecover recovered t evs
+                Just _  -> verify (Map.delete dns toRecover)
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs:365]: You are right!

File: ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 2]
```diff
@@ -320,27 +320,40 @@ mockLocalRootPeersProvider :: forall m.
...
-      void $ MonadTimer.timeout 3600 $
-        localRootPeersProvider tracer
-                               (curry toSockAddr)
-                               DNSResolver.defaultResolvConf
-                               (mockDNSActions dnsMapScriptVar
-                                               dnsTimeoutScriptVar
-                                               dnsLookupDelayScriptVar)
-                               (readTVar localRootPeersVar)
-                               resultVar
+      withAsync (updateDNSMap dnsMapScriptVar dnsMapVar) $ \_ -> do
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:333]: This change is part of a refactor to make iDomainMap mutable so I can test DNS changes.

File: ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs [hunk 3]
```diff
@@ -265,6 +273,115 @@ prop_diffusion_nolivelock defaultBearerInfo diffScript@(DiffusionScript l) =
...
+             Nothing -> verify toRecover recovered t evs
+             Just _  -> verify (Map.delete dns toRecover)
+                                (recovered + 1)
+                                t
+                                evs
+        DiffusionDiffusionSimulationTrace TrReconfigurionNode ->
+          verify Map.empty recovered t evs
+        _ -> verify toRecover recovered time evs
+
+    verify_dns_can_recover :: Events DiffusionTestTrace -> Property
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs]: Done!

File: ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs [hunk 4]
```diff
@@ -265,6 +273,115 @@ prop_diffusion_nolivelock defaultBearerInfo diffScript@(DiffusionScript l) =
...
+                     ++ "Final time: " ++ show time ++ "\n"
+                     ++ "Number of recovered: " ++ show recovered )
+                     (all (>= time) toRecover || recovered > 0)
+    verify toRecover recovered time ((t, ev):evs) =
+      case ev of
+        DiffusionLocalRootPeerTrace
+          (TraceLocalRootFailure dap (DNSError err)) ->
+            let ttl = ttlForDnsError err 0
+                dns = dapDomain dap
+             in verify (Map.singleton dns (addTime ttl t) <> toRecover)
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs]: Done!

File: ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs]: Done!

File: ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs [hunk 5]
```diff
@@ -265,6 +273,115 @@ prop_diffusion_nolivelock defaultBearerInfo diffScript@(DiffusionScript l) =
...
+           in case Map.lookup dns toRecover of
+                Nothing -> verify toRecover recovered t evs
+                Just _  -> verify (Map.delete dns toRecover)
+                                  (recovered + 1)
+                                  t
+                                  evs
+        DiffusionPublicRootPeerTrace (TracePublicRootResult dns _) ->
+           case Map.lookup dns toRecover of
+             Nothing -> verify toRecover recovered t evs
+             Just _  -> verify (Map.delete dns toRecover)
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs]: Done!

File: ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs]: Actually we shouldn't. The reason we track the `TTLs + current time` is due to us truncating the log and not seeing the result of the next attempt. The next attempt can not be made before the `TTL Difftime` so the condition you said might not be true. We could test for the negation of that condition and check that the `TTL` is being respected but we are always setting the `ttlForDNsError` to be the smallest possible (Notice: we always call `ttlDorDnsError err 0`)

**Review by coot — APPROVED** [review]
LGTM

**Comment by bolt12** [comment]
bors merge

**Pull request merged**